qiskit_nature/transformers/second_quantization/electronic/seniority_zero_transformer.py

```diff
@@ -0,0 +1,246 @@
+                    )
+                    continue
+
+            if len(transformed_property._properties) == 0:
+                transformed_property = None
+
+        elif isinstance(prop, ElectronicEnergy):
+            h_1 = prop.get_electronic_integral(ElectronicBasis.MO, 1).get_matrix()
+            h_2 = prop.get_electronic_integral(ElectronicBasis.MO, 2).get_matrix()
+
+            hr_1, hr_2 = self._restricted_electronic_integrals(h_1, h_2)
+
+            transformed_property = ElectronicEnergy.from_raw_integrals(
+                ElectronicBasis.MO, hr_1, hr_2
+            )
+            transformed_property._shift = prop._shift.copy()
+            transformed_property.nuclear_repulsion_energy = prop.nuclear_repulsion_energy
+            transformed_property.reference_energy = prop.reference_energy
+
+            transformed_property.second_q_ops = partial(  # type: ignore[assignment]
+                second_q_ops_electronic_energy, transformed_property
+            )
+
+        elif isinstance(prop, ParticleNumber):
+            transformed_property = prop
+
+            transformed_property.second_q_ops = partial(  # type: ignore[assignment]
+                second_q_ops_particle_number, transformed_property
+            )
+        elif isinstance(prop, DriverMetadata):
+            transformed_property = prop
+        elif isinstance(prop, SecondQuantizedProperty):
+            transformed_property = None
+        elif isinstance(prop, ElectronicBasisTransform):
+            transformed_property = None
+
+        else:
+            raise TypeError(f"{type(prop)} is an unsupported Property-type for this Transformer!")
+
+        return transformed_property
+
+    def _restricted_electronic_integrals(
+        self, h_1: np.ndarray, h_2: np.ndarray
+    ) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Transform integrals to the restricted formalism
+        arxiv:2002.00035, Appendix A, Equations (A10),(A11),(A12)
+        https://arxiv.org/abs/2002.00035
+
+        Args:
+            h_1 (np.ndarray): one-body integrals, (N x N)
+            h_2 (np.ndarray): two-body integrals, (N x N x N x N)
+
+        Returns:
+            hr_1 (np.ndarray): restricted one-body integrals, (N x N)
+            hr_2 (np.ndarray): restricted two-body integrals, (N x N)
+        """
+        # change of notation
+        h_2 = np.einsum("ijkl->ljik", h_2)
+
+        hr_1 = np.zeros((self._num_modes, self._num_modes))
+        hr_2 = np.zeros((self._num_modes, self._num_modes))
+
+        for i in range(self._num_modes):
+            hr_1[i, i] = 2 * h_1[i, i] + h_2[i, i, i, i]
+
+            for j in range(self._num_modes):
+                if i != j:
+                    hr_1[i, j] = h_2[i, i, j, j]
+                    hr_2[i, j] = 2 * h_2[i, j, j, i] - h_2[i, j, i, j]
+
+        return hr_1, hr_2
+
+
+def second_q_ops_electronic_energy(self: ElectronicEnergy) -> ListOrDictType[VibrationalOp]:
```

Comment:
Should this be public in this class, I think it would be better annotated as private. As a function, at first sight, having self as the first parameter is confusing, but then I see above this is monkey patched into an class instance where self is the target of the monkey patching. Since I also see that monkey patching has already been commented on, hopefully something better is figured out in general.

Reply:
What do you think about the strategy pattern suggestion? It is not perfect but the best idea I have had so far in order to remove the need for monkeypatching

Reply:
The monkey patching is really in effect just dynamically creating a subclasses of ElectronicEnergy etc. Could they be created as permanent sub-classes under the `vibrational' side and just used where the transform changes it from an ElectronicEnergy in the `electronic` side to an equivalent there with just the second_q_ops overidden (that is all I see monkey patched). Its always going to alter from one property type to another right.

Reply:
@pbark and I were discussing with @jschuhmac offline, that the naming `vibrational` is more than bad in this scenario. We are using this operator in this scenario because we want to describe _pairs of fermions_ which, together, behave like a boson. When we originally wrote the `VibrationalOp`, we voted against the name `BosonicOp` for reasons which I do not recall in all detail. However, now, I believe we need to revisit that naming choice.

All that said, I am actually wondering whether we want to make the `SeniorityZero` more of a first-class citizen by taking an approach of implementing two methods along the lines of `to_fermionic_op()` (for the current "default" behavior) and `to_vibrational_op()` (with an updated name depending on what we discuss above) in all properties directly.
Alternatively, we could even envision providing some sort of argument to `second_q_ops()` or `to_second_q_op()` (this needs to be unified, but this is another issue).
I am not sure whether this scales because I do not recall all of the details which the `ParticleHoleTransformer` will bring (i.e. does it affect only data or also the strategy) but I wanted to at least put this option up for discussion.

Reply:
I don't fully follow - as you say the transform makes the operator bosonic in nature by working with pairs, so I do not quite see how a to_fermionic_op applies. I was more thinking here that the properties were being transformed into something bosonic in nature and hence properties reflecting that so that the operator built would be bosonic. Or maybe you are thinking that properties have these methods for fermionic or bosonic and can implement one of the other as in your option to second_q_ops. Is there an option here though where some property could reasonably emit either?

My recollection of PH transform as it used to be was just a change in the data along with an energy extracted where the energy is the HF energy so that UCCSD first iteration will emit 0 once this is done.